File: benchmarks/matrix_multiply/matrix_multiply.rb
n = ARGV[0] ? ARGV[0].to_i : 200

start = Process.clock_gettime(Process::CLOCK_MONOTONIC)

# Initialize matrices
a = Array.new(n) { |i| Array.new(n) { |j| (i + j).to_f / n } }
b = Array.new(n) { |i| Array.new(n) { |j| (i - j).to_f / n } }
c = Array.new(n) { Array.new(n, 0.0) }

# Matrix multiply: C = A * B
n.times do |i|
  n.times do |j|
    sum = 0.0
    n.times do |k|
      sum += a[i][k] * b[k][j]
    end
    c[i][j] = sum
  end
end

# Compute checksum
checksum = c.flatten.sum

elapsed = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000
STDERR.puts "TIME_MS:#{format('%.2f', elapsed)}"
puts format("%.6f", checksum)

File: benchmarks/matrix_multiply/matrix_multiply.js
const n = parseInt(process.argv[2]) || 200;

// Initialize matrices
const a = [];
const b = [];
const c = [];

for (let i = 0; i < n; i++) {
    a[i] = [];
    b[i] = [];
    c[i] = [];
    for (let j = 0; j < n; j++) {
        a[i][j] = (i + j) / n;
        b[i][j] = (i - j) / n;
        c[i][j] = 0.0;
    }
}

// Matrix multiply: C = A * B
for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
        let sum = 0.0;
        for (let k = 0; k < n; k++) {
            sum += a[i][k] * b[k][j];
        }
        c[i][j] = sum;
    }
}

// Compute checksum
let checksum = 0.0;
for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
        checksum += c[i][j];
    }
}
console.log(checksum.toFixed(6));
